### ISSUE TITLE
spglib.get_symmetry_dataset cannot find the correct origin shft with realistic input

### ISSUE THREAD

**Issue by lyuwen** [atztogo/spglib#68]
This issue would appear using the following script
```import spglib
# Example 1
# Graphene realistic
a = 2.44
lattice = [[  2.1217113285639000,   1.2249706066897641,   0.0000000000000000 ],
           [  2.1217113285639000,  -1.2249706066897641,   0.0000000000000000 ],
           [  0.0000000000000000,   0.0000000000000000, -15.0000000000000000 ]]

points = [[ 0.00000000, 0.00000000, 0.00000000 ],
          [ 0.33333333, 0.33333333, 0.00000000 ]]

numbers = [6, 6]

cell=(lattice, points, numbers)
dataset = spglib.get_symmetry_dataset(cell, symprec=1e-6)

print("Space group type: %s (%d)"
      % (dataset['international'], dataset['number']))
print("Transformation matrix:")
for x in dataset['transformation_matrix']:
    print("  %2d %2d %2d" % tuple(x))
print("Origin shift: %f %f %f" % tuple(dataset['origin_shift']))


# Example 2
# Graphene
a = 2.44
lattice = [[  np.sqrt(3) / 2 * a,   0.5 * a,   0.0000000000000000 ],
           [  np.sqrt(3) / 2 * a,  -0.5 * a,   0.0000000000000000 ],
           [  0.0000000000000000,   0.0000000000000000, -15.0000000000000000 ]]

points = [[ 0.00000000, 0.00000000, 0.00000000 ],
          [ 1./3., 1./3., 0.00000000 ]]

numbers = [6, 6]

cell=(lattice, points, numbers)
dataset = spglib.get_symmetry_dataset(cell)

print("Space group type: %s (%d)"
      % (dataset['international'], dataset['number']))
print("Transformation matrix:")
for x in dataset['transformation_matrix']:
    print("  %2d %2d %2d" % tuple(x))
print("Origin shift: %f %f %f" % tuple(dataset['origin_shift']))
```

They would return different result. 

And I have found that this issue appears on and after the commit e106ae1ad3146ed47c597de450326ec2132fea00. Before this it can give the same result.

Do you have any insights on this?

What I currently know is that in `src/refinement.c` method `find_similar_bravais_lattice`. There is a piece of code trying to find the shortest origin shift. Before that the origin shift vector of example 1 has a value that agrees with example 2, but after that the vector example 1 becomes 0.

**Comment by atztogo** [comment]
Thanks. I wii have a look.

**Comment by lyuwen** [comment]
Should it be 
``` 
p[j] -= tmp_vec[j]
``` 
instead of
``` 
p[j] += tmp_vec[j]
``` 
In line line 1412 (file `src/refinement.c`) ? This would fix this issue, but I’m not sure since I don’t have any unit tests.

**Comment by atztogo** [comment]
I don't see why you think so. To understand, I need your formulation. But this is not the point we should discuss.

I think the code is OK. The point is that the convention to choose origin shift was changed. So the behaviour changed.


```
lattice = [[  2.1217113285639000,   1.2249706066897641,   0.0000000000000000 ],
           [  2.1217113285639000,  -1.2249706066897641,   0.0000000000000000 ],
           [  0.0000000000000000,   0.0000000000000000, -15.0000000000000000 ]]
points = [[ 0.00000000, 0.00000000, 0.00000000 ],
          [ 0.33333333, 0.33333333, 0.00000000 ]]
numbers = [6, 6]

cell=(lattice, points, numbers)
dataset = spglib.get_symmetry_dataset(cell, symprec=1e-6)

print("Space group type: %s (%d)"
      % (dataset['international'], dataset['number']))
print("Transformation matrix:")
for x in dataset['transformation_matrix']:
    print("  %2f %2f %2f" % tuple(x))
print("Origin shift: %f %f %f" % tuple(dataset['origin_shift']))
print("Standardized structure")
print(dataset['std_lattice'])
print(dataset['std_positions'])

print("")

# Example 2
# Graphene
a = 2.44
lattice = [[  np.sqrt(3) / 2 * a,   0.5 * a,   0.0000000000000000 ],
           [  np.sqrt(3) / 2 * a,  -0.5 * a,   0.0000000000000000 ],
           [  0.0000000000000000,   0.0000000000000000, -15.0000000000000000 ]]
points = [[ 0.00000000, 0.00000000, 0.00000000 ],
          [ 1./3., 1./3., 0.00000000 ]]
numbers = [6, 6]

cell=(lattice, points, numbers)
dataset = spglib.get_symmetry_dataset(cell)

print("Space group type: %s (%d)"
      % (dataset['international'], dataset['number']))
print("Transformation matrix:")
for x in dataset['transformation_matrix']:
    print("  %2f %2f %2f" % tuple(x))
print("Origin shift: %f %f %f" % tuple(dataset['origin_shift']))

print("Standardized structure")
print(dataset['std_lattice'])
print(dataset['std_positions'])
```

This gives

```
Space group type: P6/mmm (191)
Transformation matrix:
  1.000000 0.000000 0.000000
  0.000000 -1.000000 0.000000
  0.000000 0.000000 -1.000000
Origin shift: 0.000000 0.000000 0.000000
Standardized structure
[[ 2.44994121  0.          0.        ]
 [-1.22497061  2.12171133  0.        ]
 [ 0.          0.         15.        ]]
[[0.         0.         0.        ]
 [0.33333333 0.66666667 0.        ]]

Space group type: P6/mmm (191)
Transformation matrix:
  1.000000 1.000000 0.000000
  1.000000 0.000000 0.000000
  0.000000 0.000000 -1.000000
Origin shift: 0.666667 0.333333 0.000000
Standardized structure
[[ 2.44        0.          0.        ]
 [-1.22        2.11310199  0.        ]
 [ 0.          0.         15.        ]]
[[0.66666667 0.33333333 0.        ]
 [0.33333333 0.66666667 0.        ]]
```

Because of slight difference of basis vectors, the transformation matrix was chosen differently, but these are equivalent in symmetry sense. The difference of transformation matrix gives different origin shifts in this case.

Your choice of basis vectors does not agree with spglib convention. If you choose the one of spglib, the transformation matrix will be identity matrix. So this problem will not happen. If this happens again, probably spglib has a bug.

**Comment by lyuwen** [comment]
Yes. But the first part sets the origin at one of the basis atom. And the second sets it at the origin of the point symmetry. Do you mean these are equivalent?

**Comment by atztogo** [comment]
I don't understand what you mean.

**Comment by atztogo** [comment]
If you want to change the bahaviour as you expect, probably you can tune this, https://github.com/atztogo/spglib/blob/master/src/delaunay.c . In spglib, the algorithm first searches primitive basis vectors, because 1. the input may not be a primitive cell, 2. the input is a primitive cell, but the choice of basis vectors is very unusual. For this, a choice of primitive basis vectors is searched by Delaunay reduction. Small difference of input basis vectors values probably results in different primitive basis vectors in this part. Maybe we can tune this code to be more mild to give similar choices for similar inputs, though I don't guarantee it.

It is very difficult to make spglib to satisfy everybody who has each expecting behaviour though they are symmetrically equivalent, but their conventions are difference.

**Comment by lyuwen** [comment]
```
Space group type: P6/mmm (191)
Transformation matrix:
  1.000000 0.000000 0.000000
  0.000000 -1.000000 0.000000
  0.000000 0.000000 -1.000000
Origin shift: 0.000000 0.000000 0.000000
Standardized structure
[[ 2.44994121  0.          0.        ]
 [-1.22497061  2.12171133  0.        ]
 [ 0.          0.         15.        ]]
[[0.         0.         0.        ]
 [0.33333333 0.66666667 0.        ]]
```
In this case the code is claiming the wyckoff position to be a and c for the two atoms. But the two atoms can only be at c position as they only have -6m2 site symmetry.

**Comment by atztogo** [comment]
I don't yet get what you mean. You may give an important point to find a possible bug. Can you explain more in details?

**Comment by lyuwen** [comment]
What I mean is that from the above structure (copied below)
```
lattice = [[  2.1217113285639000,   1.2249706066897641,   0.0000000000000000 ],
           [  2.1217113285639000,  -1.2249706066897641,   0.0000000000000000 ],
           [  0.0000000000000000,   0.0000000000000000, -15.0000000000000000 ]]
points = [[ 0.00000000, 0.00000000, 0.00000000 ],
          [ 0.33333333, 0.33333333, 0.00000000 ]]
```
visualized as `case 1` in the below picture (where star is the origin i.e. coordinate (0,0)):
![figure_1-1](https://user-images.githubusercontent.com/17034696/49488390-76be0880-f814-11e8-92bd-4ef7aa74ce91.png)
we know that it has space group  P6/mmm (191), and we can also easily find out that the site symmetry on each atom is -6m2(D3h).
So for this structure the Wyckoff Letters of the two atoms must be **c** (from [here](http://www.cryst.ehu.es/cgi-bin/cryst/programs/nph-wp-list?gnum=191), or [international tables of crystalgraphy](https://it.iucr.org/Ac/ch2o3v0001/sgtable2o3o191/)), as it is a 2-D material.
So the **standard** structure in my mind should follow this international table, which would put the atoms at (1/3,2/3,0) and (2/3,1/3,0). which will result in the `case 2` points in the picture above.
As of now the code is claiming the wyckoff letters to be a and c for the two atoms and put the atoms at (0, 0, 0) and (1/3,2/3,0).

**Comment by atztogo** [comment]
Have you seen `print(dataset['wyckoffs'])`? For me, it gives c for both. If not, could you give the code example?

**Comment by lyuwen** [comment]
```
import spglib

lattice = [[  2.1217113285639000,   1.2249706066897641,   0.0000000000000000 ],
           [  2.1217113285639000,  -1.2249706066897641,   0.0000000000000000 ],
           [  0.0000000000000000,   0.0000000000000000, -15.0000000000000000 ]]
points = [[ 0.00000000, 0.00000000, 0.00000000 ],
          [ 0.33333333, 0.33333333, 0.00000000 ]]
numbers = [6, 6]

cell=(lattice, points, numbers)
dataset = spglib.get_symmetry_dataset(cell, symprec=1e-6)

print("Space group type: %s (%d)"
      % (dataset['international'], dataset['number']))
print("Transformation matrix:")
for x in dataset['transformation_matrix']:
    print("  %2f %2f %2f" % tuple(x))
print("Origin shift: %f %f %f" % tuple(dataset['origin_shift']))
print("Standardized structure")
print(dataset['std_lattice'])
print(dataset['std_positions'])
print(dataset['wyckoffs'])

print("")

# Example 2
# Graphene
a = 2.44
lattice = [[  np.sqrt(3) / 2 * a,   0.5 * a,   0.0000000000000000 ],
           [  np.sqrt(3) / 2 * a,  -0.5 * a,   0.0000000000000000 ],
           [  0.0000000000000000,   0.0000000000000000, -15.0000000000000000 ]]
points = [[ 0.00000000, 0.00000000, 0.00000000 ],
          [ 1./3., 1./3., 0.00000000 ]]
numbers = [6, 6]

cell=(lattice, points, numbers)
dataset = spglib.get_symmetry_dataset(cell)

print("Space group type: %s (%d)"
      % (dataset['international'], dataset['number']))
print("Transformation matrix:")
for x in dataset['transformation_matrix']:
    print("  %2f %2f %2f" % tuple(x))
print("Origin shift: %f %f %f" % tuple(dataset['origin_shift']))

print("Standardized structure")
print(dataset['std_lattice'])
print(dataset['std_positions'])
print(dataset['wyckoffs'])
```
gives
```
Space group type: P6/mmm (191)
Transformation matrix:
  1.000000 0.000000 0.000000
  0.000000 -1.000000 0.000000
  0.000000 0.000000 -1.000000
Origin shift: 0.000000 0.000000 0.000000
Standardized structure
[[ 2.44994121  0.          0.        ]
 [-1.22497061  2.12171133  0.        ]
 [ 0.          0.         15.        ]]
[[0.         0.         0.        ]
 [0.33333333 0.66666667 0.        ]]
['a', 'c']

Space group type: P6/mmm (191)
Transformation matrix:
  1.000000 1.000000 0.000000
  1.000000 0.000000 0.000000
  0.000000 0.000000 -1.000000
Origin shift: 0.666667 0.333333 0.000000
Standardized structure
[[ 2.44        0.          0.        ]
 [-1.22        2.11310199  0.        ]
 [ 0.          0.         15.        ]]
[[0.66666667 0.33333333 0.        ]
 [0.33333333 0.66666667 0.        ]]
['c', 'c']
```

**Comment by lyuwen** [comment]
Case 2 gives c for both. But case 1 is giving a and c for the two atoms. Note that these two structure are only slightly different, i.e. structure in case 2 is numerically perfect hexagonal, that in case 1 has small numerical error which should pass the tolerance.

**Comment by atztogo** [comment]
Which version do you use? And can you give your full script to reproduce?

**Comment by lyuwen** [comment]
version: spglib==1.11.1.post3

[test1.py.zip](https://github.com/atztogo/spglib/files/2647237/test1.py.zip)

**Comment by atztogo** [comment]
And what is the output? Mine is

```
% python test1.py
case 1
Space group type: P6/mmm (191)
Transformation matrix:
  1.000000 0.000000 0.000000
  0.000000 -1.000000 0.000000
  0.000000 0.000000 -1.000000
Origin shift: 0.333333 0.666667 0.000000
Standardized structure
[[ 2.44994121  0.          0.        ]
 [-1.22497061  2.12171133  0.        ]
 [ 0.          0.         15.        ]]
[[0.33333333 0.66666667 0.        ]
 [0.66666667 0.33333333 0.        ]]
['c', 'c']

case 2
Space group type: P6/mmm (191)
Transformation matrix:
  1.000000 1.000000 0.000000
  1.000000 0.000000 0.000000
  0.000000 0.000000 -1.000000
Origin shift: 0.666667 0.333333 0.000000
Standardized structure
[[ 2.44        0.          0.        ]
 [-1.22        2.11310199  0.        ]
 [ 0.          0.         15.        ]]
[[0.66666667 0.33333333 0.        ]
 [0.33333333 0.66666667 0.        ]]
['c', 'c']
```

If we don't agree, you can try develop branch.

**Comment by lyuwen** [comment]
Mine is exactly as I posted.

**Comment by atztogo** [comment]
Oh, true. I recompile my spglib of develop, and I get `['a', 'c']`. It should be a bug. Thanks.

**Comment by lyuwen** [comment]
> ```
> print("case 1")
> lattice = [[  2.1217113285639000,   1.2249706066897641,   0.0000000000000000 ],
>            [  2.1217113285639000,  -1.2249706066897641,   0.0000000000000000 ],
>            [  0.0000000000000000,   0.0000000000000000, -15.0000000000000000 ]]
> points = [[ 0.00000000, 0.00000000, 0.00000000 ],
>           [ 0.33333333, 0.33333333, 0.00000000 ]]
> numbers = [6, 6]
> 
> cell=(lattice, points, numbers)
> dataset = spglib.get_symmetry_dataset(cell, symprec=1e-6)
> 
> print("Space group type: %s (%d)"
>       % (dataset['international'], dataset['number']))
> print("Transformation matrix:")
> for x in dataset['transformation_matrix']:
>     print("  %2f %2f %2f" % tuple(x))
> print("Origin shift: %f %f %f" % tuple(dataset['origin_shift']))
> print("Standardized structure")
> print(dataset['std_lattice'])
> print(dataset['std_positions'])
> print(dataset['wyckoffs'])
> 
> print("")
> 
> print("case 2")
> a = 2.44
> lattice = [[  np.sqrt(3) / 2 * a,   0.5 * a,   0.0000000000000000 ],
>            [  np.sqrt(3) / 2 * a,  -0.5 * a,   0.0000000000000000 ],
>            [  0.0000000000000000,   0.0000000000000000, -15.0000000000000000 ]]
> points = [[ 0.00000000, 0.00000000, 0.00000000 ],
>           [ 1./3., 1./3., 0.00000000 ]]
> numbers = [6, 6]
> 
> cell=(lattice, points, numbers)
> dataset = spglib.get_symmetry_dataset(cell)
> 
> print("Space group type: %s (%d)"
>       % (dataset['international'], dataset['number']))
> print("Transformation matrix:")
> for x in dataset['transformation_matrix']:
>     print("  %2f %2f %2f" % tuple(x))
> print("Origin shift: %f %f %f" % tuple(dataset['origin_shift']))
> 
> print("Standardized structure")
> print(dataset['std_lattice'])
> print(dataset['std_positions'])
> print(dataset['wyckoffs'])
> ```
> gives
> 
> ```
> case 1
> Space group type: P6/mmm (191)
> Transformation matrix:
>   1.000000 0.000000 0.000000
>   0.000000 -1.000000 0.000000
>   0.000000 0.000000 -1.000000
> Origin shift: 0.000000 0.000000 0.000000
> Standardized structure
> [[ 2.44994121  0.          0.        ]
>  [-1.22497061  2.12171133  0.        ]
>  [ 0.          0.         15.        ]]
> [[0.         0.         0.        ]
>  [0.33333333 0.66666667 0.        ]]
> ['a', 'c']
> 
> case 2
> Space group type: P6/mmm (191)
> Transformation matrix:
>   1.000000 1.000000 0.000000
>   1.000000 0.000000 0.000000
>   0.000000 0.000000 -1.000000
> Origin shift: 0.666667 0.333333 0.000000
> Standardized structure
> [[ 2.44        0.          0.        ]
>  [-1.22        2.11310199  0.        ]
>  [ 0.          0.         15.        ]]
> [[0.66666667 0.33333333 0.        ]
>  [0.33333333 0.66666667 0.        ]]
> ['c', 'c']
> ```

From the output, I think the definition of the p vector is not agreeing with that in the [documentation](https://atztogo.github.io/spglib/definition.html#transformation-matrix-boldsymbol-p-and-origin-shift-boldsymbol-p).

**Comment by atztogo** [comment]
Yes, you are correct. But I do something a bit tricky to rotate crystal. Probably my formulation or implementation is incorrect. Now I am checking. Thanks for your effort to survey.

**Comment by atztogo** [comment]
I finally found the bug in my formulation. I think I could fix it and now it is in develop branch. I need to write some more tests such as also checking Wyckoff positions because the previous version unfortunately could pass my test. I pushed myself very hard for these two days and it was a good stress test of my brain. Thanks for your being patient to tell the bug to me.

**Comment by lyuwen** [comment]
Great. It seems the problem with the structure I posted is fixed. Thank you for the update. This is indeed a very tricky problem.

**Comment by atztogo** [comment]
I released the bug fixed version v1.11.2. I will close this issue.
